pkg/controllermanager/controller/shootstate/add.go

```diff
@@ -0,0 +1,117 @@
+	if r.Client == nil {
+		r.Client = mgr.GetClient()
+	}
+
+	return builder.ControllerManagedBy(mgr).
+		Named(ControllerName).
+		For(
+			&gardencorev1beta1.ShootState{},
+			builder.WithPredicates(predicateutils.ForEventTypes(predicateutils.Create, predicateutils.Update)),
+		).
+		WithOptions(controller.Options{
+			MaxConcurrentReconciles: ptr.Deref(r.Config.ConcurrentSyncs, 0),
+		}).
+		Watches(
+			&gardencorev1beta1.Shoot{},
+			handler.EnqueueRequestsFromMapFunc(r.MapShootToShootState),
+			builder.WithPredicates(r.ShootPredicates()),
+		).Complete(r)
+}
+
+// MapShootToShootState maps a Shoot object to ShootState reconciliation request.
+func (r *Reconciler) MapShootToShootState(_ context.Context, obj client.Object) []reconcile.Request {
+	shoot, ok := obj.(*gardencorev1beta1.Shoot)
+	if !ok {
+		return nil
+	}
+
+	namespacedName := types.NamespacedName{
+		Name:      shoot.Name,
+		Namespace: shoot.Namespace,
+	}
+	return []reconcile.Request{{NamespacedName: namespacedName}}
+}
+
+// ShootPredicates returns predicates for Shoot requests acceptance.
+func (r *Reconciler) ShootPredicates() predicate.Predicate {
+	return predicate.Funcs{
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			shoot, ok := e.ObjectNew.(*gardencorev1beta1.Shoot)
+			if !ok || shoot.Status.LastOperation == nil {
+				return false
+			}
+
+			shootOld, ok := e.ObjectOld.(*gardencorev1beta1.Shoot)
+			if !ok || shootOld.Status.LastOperation == nil {
+				return false
+			}
+
+			var (
+				oldLastOpType  = shootOld.Status.LastOperation.Type
+				newLastOpType  = shoot.Status.LastOperation.Type
+				newLastOpState = shoot.Status.LastOperation.State
+
+				isOldLastOpReconcile = oldLastOpType == gardencorev1beta1.LastOperationTypeReconcile
+				isOldLastOpMigrate   = oldLastOpType == gardencorev1beta1.LastOperationTypeMigrate
+				isOldLastOpRestore   = oldLastOpType == gardencorev1beta1.LastOperationTypeRestore
+
+				isNewLastOpReconcile = newLastOpType == gardencorev1beta1.LastOperationTypeReconcile
+				isNewLastOpMigrate   = newLastOpType == gardencorev1beta1.LastOperationTypeMigrate
+				isNewLastOpRestore   = newLastOpType == gardencorev1beta1.LastOperationTypeRestore
+				isNewLastOpSucceeded = newLastOpState == gardencorev1beta1.LastOperationStateSucceeded
+			)
+
+			isMigrating := false
+			if isOldLastOpReconcile && isNewLastOpMigrate {
+				// Shoot last operation gets updated from Reconcile to Migrate type.
+				isMigrating = true
+			} else if isOldLastOpMigrate && isNewLastOpRestore {
+				// Shoot last operation gets updated from Migrate to Restore type.
+				isMigrating = true
+			} else if isOldLastOpRestore && isNewLastOpRestore && isNewLastOpSucceeded {
+				// Shoot last operation gets updated from Restore to Restore type with Succeeded state.
+				isMigrating = true
+			} else if isOldLastOpRestore && isNewLastOpReconcile {
+				// Shoot last operation gets updated from Restore to Reconcile.
+				isMigrating = true
```

Comment:
`isMigrating` is a bit confusing as a variable name. This and the previous case the shoot migration has finished successfully, so the shoot is no longer being migration. Perhaps something like:
```suggestion
				enqueueShootState = true
```

Reply:
Thanks for the suggestion, renamed the variable to `enqueueShootState` in my latest updates.